File: packages/sdk-ts/src/adapters/tavily.ts
// Authoritative one-credit Tavily Search adapter.
//
// This adapter intentionally accepts only basic search with automatic
// parameters disabled. Advanced/automatic search must use controlledUsage
// with an explicit two-credit maximum and provider usage extractor.

import { isProxy } from 'node:util/types';
import { controlledUsage, type ControlledUsageResult } from '../core/controlled_usage.js';
import { linkLocalControlledCallbackNoDispatch } from '../core/control_correlation.js';
import { PylvaControlValidationError } from '../errors/control.js';

export const TAVILY_SEARCH_COST_SOURCE_SLUG = 'tavily-search' as const;
export const TAVILY_SEARCH_TOOL_NAME = 'Tavily Search' as const;
export const TAVILY_SEARCH_METRIC = 'credit' as const;
export const TAVILY_BASIC_SEARCH_CREDITS = '1' as const;

export interface TavilySearchClient<T> {
  search(query: string, options?: Readonly<Record<string, unknown>>): Promise<T>;
}

export interface ControlledTavilySearchInput {
  query: string;
  customerId?: string;
  step?: string;
  searchDepth?: 'basic' | 'advanced' | 'fast' | 'ultra-fast';
  autoParameters?: boolean;
  searchOptions?: Readonly<Record<string, unknown>>;
  reservationTtlSeconds?: number;
  heartbeatIntervalMs?: number;
}

const INPUT_KEYS = new Set([
  'query',
  'customerId',
  'step',
  'searchDepth',
  'autoParameters',
  'searchOptions',
  'reservationTtlSeconds',
  'heartbeatIntervalMs',
]);
const MAX_OPTION_DEPTH = 8;
const MAX_OPTION_NODES = 1_024;
const MAX_OPTION_KEYS = 128;
const MAX_OPTION_KEY_LENGTH = 128;
const MAX_OPTION_ARRAY_LENGTH = 256;
const MAX_OPTION_STRING_LENGTH = 16_384;

function validationError(): PylvaControlValidationError {
  linkLocalControlledCallbackNoDispatch('tool');
  return new PylvaControlValidationError('controlledTavilySearch');
}

function defineData(target: Record<string, unknown>, key: string, value: unknown): void {
  Object.defineProperty(target, key, {
    value,
    enumerable: true,
    writable: true,
    configurable: true,
  });
}

function strictInput(value: unknown): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value) || isProxy(value)) {
    throw validationError();
  }
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) {
    throw validationError();
  }
  const snapshot = Object.create(null) as Record<string, unknown>;
  for (const key of Reflect.ownKeys(value)) {
    if (typeof key !== 'string' || key === '__proto__' || !INPUT_KEYS.has(key)) {
      throw validationError();
    }
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor === undefined || !descriptor.enumerable || !('value' in descriptor)) {
      throw validationError();
    }
    defineData(snapshot, key, descriptor.value);
  }
  return snapshot;
}

interface SnapshotState {
  readonly seen: WeakSet<object>;
  nodes: number;
}

function snapshotPlainJson(value: unknown, state: SnapshotState, depth = 0): unknown {
  state.nodes += 1;
  if (state.nodes > MAX_OPTION_NODES || depth > MAX_OPTION_DEPTH) throw validationError();
  if (value === null || typeof value === 'boolean') return value;
  if (typeof value === 'string') {
    if (value.length > MAX_OPTION_STRING_LENGTH) throw validationError();
    return value;
  }
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) throw validationError();
    return value;
  }
  if (typeof value !== 'object' || isProxy(value)) throw validationError();
  if (state.seen.has(value)) throw validationError();
  state.seen.add(value);

  if (Array.isArray(value)) {
    if (
      Object.getPrototypeOf(value) !== Array.prototype ||
      value.length > MAX_OPTION_ARRAY_LENGTH
    ) {
      throw validationError();
    }
    let indexedKeys = 0;
    for (const key of Reflect.ownKeys(value)) {
      if (key === 'length') continue;
      if (typeof key !== 'string' || !/^(?:0|[1-9][0-9]*)$/.test(key)) {
        throw validationError();
      }
      const index = Number(key);
      if (!Number.isSafeInteger(index) || index >= value.length || String(index) !== key) {
        throw validationError();
      }
      indexedKeys += 1;
    }
    if (indexedKeys !== value.length) throw validationError();
    const snapshot: unknown[] = [];
    for (let index = 0; index < value.length; index += 1) {
      const descriptor = Object.getOwnPropertyDescriptor(value, String(index));
      if (descriptor === undefined || !descriptor.enumerable || !('value' in descriptor)) {
        throw validationError();
      }
      snapshot.push(snapshotPlainJson(descriptor.value, state, depth + 1));
    }
    return snapshot;
  }

  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) throw validationError();
  const keys = Reflect.ownKeys(value);
  if (keys.length > MAX_OPTION_KEYS) throw validationError();
  const snapshot = Object.create(null) as Record<string, unknown>;
  for (const key of keys) {
    if (typeof key !== 'string' || key === '__proto__' || key.length > MAX_OPTION_KEY_LENGTH) {
      throw validationError();
    }
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor === undefined || !descriptor.enumerable || !('value' in descriptor)) {
      throw validationError();
    }
    defineData(snapshot, key, snapshotPlainJson(descriptor.value, state, depth + 1));
  }
  return snapshot;
}

function searchOptions(
  raw: unknown,
  searchDepthInput: unknown,
  autoParametersInput: unknown,
): Readonly<Record<string, unknown>> {
  const searchDepth = searchDepthInput === undefined ? 'basic' : searchDepthInput;
  const autoParameters = autoParametersInput === undefined ? false : autoParametersInput;
  if (searchDepth !== 'basic' || autoParameters !== false) {
    throw validationError();
  }
  const options = snapshotPlainJson(raw === undefined ? {} : raw, {
    seen: new WeakSet(),
    nodes: 0,
  });
  if (typeof options !== 'object' || options === null || Array.isArray(options)) {
    throw validationError();
  }
  const optionRecord = options as Record<string, unknown>;
  // @tavily/core forwards unknown snake_case keys after its mapped fields, so
  // aliases of locked fields must be rejected instead of merely overwritten.
  if (
    Object.hasOwn(optionRecord, 'query') ||
    Object.hasOwn(optionRecord, 'search_depth') ||
    Object.hasOwn(optionRecord, 'auto_parameters') ||
    Object.hasOwn(optionRecord, 'include_usage') ||
    (optionRecord['searchDepth'] ?? 'basic') !== 'basic' ||
    (optionRecord['autoParameters'] ?? false) !== false
  ) {
    throw validationError();
  }
  // Caller values cannot weaken the exact one-credit contract.
  defineData(optionRecord, 'searchDepth', 'basic');
  defineData(optionRecord, 'autoParameters', false);
  defineData(optionRecord, 'includeUsage', true);
  return optionRecord;
}

function boundSearch<T>(client: TavilySearchClient<T>): TavilySearchClient<T>['search'] {
  if (typeof client !== 'object' || client === null || isProxy(client)) throw validationError();
  let search: unknown;
  let owner: object | null = client;
  let traversed = 0;
  while (owner !== null && traversed <= 16) {
    if (isProxy(owner)) throw validationError();
    const descriptor = Object.getOwnPropertyDescriptor(owner, 'search');
    if (descriptor !== undefined) {
      if (!('value' in descriptor)) throw validationError();
      search = descriptor.value;
      break;
    }
    owner = Object.getPrototypeOf(owner) as object | null;
    traversed += 1;
  }
  if (typeof search !== 'function' || isProxy(search)) throw validationError();
  return (query, options) => Reflect.apply(search, client, [query, options]) as Promise<T>;
}

function credits(response: unknown): string {
  if (typeof response !== 'object' || response === null || Array.isArray(response)) {
    throw new TypeError('missing Tavily usage');
  }
  const usage = (response as Record<string, unknown>)['usage'];
  if (typeof usage !== 'object' || usage === null || Array.isArray(usage)) {
    throw new TypeError('missing Tavily usage');
  }
  const value = (usage as Record<string, unknown>)['credits'];
  if (typeof value === 'number' && Number.isSafeInteger(value) && value > 0) {
    return String(value);
  }
  if (typeof value === 'string' && /^[0-9]+$/.test(value)) {
    const normalized = value.replace(/^0+(?=\d)/, '');
    if (normalized !== '0') return normalized;
  }
  throw new TypeError('missing Tavily usage');
}

/** Call `@tavily/core` search under an exact one-credit reservation. */
export async function controlledTavilySearch<T>(
  client: TavilySearchClient<T>,
  rawInput: ControlledTavilySearchInput,
): Promise<ControlledUsageResult<T>> {
  const input = strictInput(rawInput);
  // Snapshot the complete adapter input before any user-controlled transport
  // lookup and before controlledUsage can await its reservation.
  const query = input['query'];
  const customerId = input['customerId'];
  const step = input['step'];
  const reservationTtlSeconds = input['reservationTtlSeconds'];
  const heartbeatIntervalMs = input['heartbeatIntervalMs'];
  const options = searchOptions(
    input['searchOptions'],
    input['searchDepth'],
    input['autoParameters'],
  );
  if (typeof query !== 'string' || query.length === 0) {
    throw validationError();
  }
  const search = boundSearch(client);
  return controlledUsage({
    costSourceSlug: TAVILY_SEARCH_COST_SOURCE_SLUG,
    toolName: TAVILY_SEARCH_TOOL_NAME,
    metric: TAVILY_SEARCH_METRIC,
    maximumValue: TAVILY_BASIC_SEARCH_CREDITS,
    invoke: () => search(query, options),
    extractActual: credits,
    customerId: customerId as string | undefined,
    step: step as string | undefined,
    reservationTtlSeconds: reservationTtlSeconds as number | undefined,
    heartbeatIntervalMs: heartbeatIntervalMs as number | undefined,
  });
}

File: packages/sdk-py/pylva/adapters/tavily.py
"""Authoritative one-credit Tavily Search adapter.

Only ``search_depth="basic"`` with ``auto_parameters=False`` is accepted by
this adapter.  Advanced or automatic search belongs on the generic bounded
helper with an explicit two-credit maximum and a response usage extractor.
"""

from __future__ import annotations

import inspect
import math
from collections.abc import Awaitable, Callable, Mapping
from types import FunctionType, MethodType
from typing import Any, Protocol, TypeVar, cast

from ..core.control_ownership import _controlled_local_no_dispatch
from ..core.controlled_usage import (
    ControlledUsageResult,
    controlled_usage,
    controlled_usage_sync,
)
from ..errors.control import PylvaControlValidationError

T = TypeVar("T")
T_co = TypeVar("T_co", covariant=True)

TAVILY_SEARCH_COST_SOURCE_SLUG = "tavily-search"
TAVILY_SEARCH_TOOL_NAME = "Tavily Search"
TAVILY_SEARCH_METRIC = "credit"
TAVILY_BASIC_SEARCH_CREDITS = "1"
_MAX_OPTION_DEPTH = 8
_MAX_OPTION_NODES = 1_024
_MAX_OPTION_KEYS = 128
_MAX_OPTION_KEY_LENGTH = 128
_MAX_OPTION_ARRAY_LENGTH = 256
_MAX_OPTION_STRING_LENGTH = 16_384
_MAX_SAFE_INTEGER = 9_007_199_254_740_991
_MISSING = object()


def _validation_error() -> PylvaControlValidationError:
    _controlled_local_no_dispatch("tool")
    return PylvaControlValidationError("controlled_tavily_search")


class TavilySyncSearchClient(Protocol[T_co]):
    def search(self, query: str, **kwargs: Any) -> T_co: ...


class TavilyAsyncSearchClient(Protocol[T_co]):
    def search(self, query: str, **kwargs: Any) -> Awaitable[T_co]: ...


def _snapshot_plain_json(
    value: object,
    *,
    seen: set[int],
    nodes: list[int],
    depth: int = 0,
) -> object:
    """Copy one bounded tree without invoking custom container hooks."""

    nodes[0] += 1
    if nodes[0] > _MAX_OPTION_NODES or depth > _MAX_OPTION_DEPTH:
        raise _validation_error()
    value_type = type(value)
    if value is None or value_type is bool:
        return value
    if value_type is str:
        if len(cast(str, value)) > _MAX_OPTION_STRING_LENGTH:
            raise _validation_error()
        return value
    if value_type is int:
        if abs(cast(int, value)) > _MAX_SAFE_INTEGER:
            raise _validation_error()
        return value
    if value_type is float:
        if not math.isfinite(cast(float, value)):
            raise _validation_error()
        return value
    if value_type not in {dict, list}:
        raise _validation_error()

    identity = id(value)
    if identity in seen:
        raise _validation_error()
    seen.add(identity)
    if value_type is list:
        source = cast(list[object], value)
        if len(source) > _MAX_OPTION_ARRAY_LENGTH:
            raise _validation_error()
        return [
            _snapshot_plain_json(item, seen=seen, nodes=nodes, depth=depth + 1) for item in source
        ]

    source_record = cast(dict[object, object], value)
    if len(source_record) > _MAX_OPTION_KEYS:
        raise _validation_error()
    snapshot: dict[str, object] = {}
    for key, item in source_record.items():
        if type(key) is not str or key == "__proto__" or len(key) > _MAX_OPTION_KEY_LENGTH:
            raise _validation_error()
        snapshot[key] = _snapshot_plain_json(
            item,
            seen=seen,
            nodes=nodes,
            depth=depth + 1,
        )
    return snapshot


def _options(
    search_options: Mapping[str, Any] | None,
    *,
    search_depth: str,
    auto_parameters: bool,
) -> dict[str, Any]:
    if search_depth != "basic" or auto_parameters is not False:
        raise _validation_error()
    # Do not iterate arbitrary Mapping implementations or subclasses: their
    # hooks can execute user code during the authoritative validation boundary.
    if search_options is not None and type(search_options) is not dict:
        raise _validation_error()
    detached = _snapshot_plain_json(
        search_options if search_options is not None else {},
        seen=set(),
        nodes=[0],
    )
    if type(detached) is not dict:  # pragma: no cover - defensive narrowing
        raise _validation_error()
    options = cast(dict[str, Any], detached)
    supplied_depth = options.get("search_depth", "basic")
    supplied_auto = options.get("auto_parameters", False)
    # Reject JavaScript aliases as well: structural/custom clients may forward
    # unknown keys after their native fields and thereby override the bound.
    if (
        supplied_depth != "basic"
        or supplied_auto is not False
        or "query" in options
        or "searchDepth" in options
        or "autoParameters" in options
        or "includeUsage" in options
    ):
        raise _validation_error()
    # These values make the one-credit bound explicit and provide exact usage
    # evidence. Caller values cannot weaken them.
    options["search_depth"] = "basic"
    options["auto_parameters"] = False
    options["include_usage"] = True
    return options


def _search_method(client: object) -> Callable[..., Any]:
    """Bind the validated transport before reservation can yield."""

    try:
        static_search = inspect.getattr_static(client, "search")
    except Exception as error:
        raise _validation_error() from error
    if type(static_search) is MethodType:
        return cast(Callable[..., Any], static_search)
    if type(static_search) is not FunctionType:
        # Properties and custom descriptors/callables can execute hooks during
        # validation or change behavior across the reservation boundary.
        raise _validation_error()
    try:
        instance_state = object.__getattribute__(client, "__dict__")
    except Exception:
        instance_state = None
    if (
        type(instance_state) is dict
        and cast(dict[object, object], instance_state).get("search", _MISSING) is static_search
    ):
        # A plain function stored directly on the instance is already unbound.
        return cast(Callable[..., Any], static_search)
    return cast(Callable[..., Any], MethodType(static_search, client))


def _credits(response: object) -> str:
    try:
        if not isinstance(response, Mapping):
            raise TypeError
        usage = response.get("usage")
        if not isinstance(usage, Mapping):
            raise TypeError
        credits = usage.get("credits")
        if isinstance(credits, bool):
            raise TypeError
        if isinstance(credits, float):
            # Provider JSON decoded as a binary float cannot participate in
            # authoritative quantity hashing, even when mathematically integral.
            raise TypeError
        if isinstance(credits, int):
            if credits <= 0:
                raise TypeError
            return str(credits)
        if isinstance(credits, str) and credits.isascii() and credits.isdigit():
            normalized = str(int(credits))
            if normalized != "0":
                return normalized
    except Exception as error:
        raise ValueError("Tavily response did not contain exact positive credit usage") from error
    raise ValueError("Tavily response did not contain exact positive credit usage")


def controlled_tavily_search_sync(
    client: TavilySyncSearchClient[T],
    query: str,
    *,
    customer_id: str | None = None,
    step: str | None = None,
    search_depth: str = "basic",
    auto_parameters: bool = False,
    search_options: Mapping[str, Any] | None = None,
    reservation_ttl_seconds: int = 300,
    heartbeat_interval_seconds: float | None = None,
) -> ControlledUsageResult[T]:
    """Call ``TavilyClient.search`` with an exact one-credit reservation."""

    if type(query) is not str or not query:
        raise _validation_error()
    options = _options(
        search_options,
        search_depth=search_depth,
        auto_parameters=auto_parameters,
    )
    search = _search_method(client)
    return controlled_usage_sync(
        cost_source_slug=TAVILY_SEARCH_COST_SOURCE_SLUG,
        tool_name=TAVILY_SEARCH_TOOL_NAME,
        metric=TAVILY_SEARCH_METRIC,
        maximum_value=TAVILY_BASIC_SEARCH_CREDITS,
        invoke=lambda: search(query, **options),
        extract_actual=_credits,
        customer_id=customer_id,
        step=step,
        reservation_ttl_seconds=reservation_ttl_seconds,
        heartbeat_interval_seconds=heartbeat_interval_seconds,
    )


async def controlled_tavily_search(
    client: TavilyAsyncSearchClient[T],
    query: str,
    *,
    customer_id: str | None = None,
    step: str | None = None,
    search_depth: str = "basic",
    auto_parameters: bool = False,
    search_options: Mapping[str, Any] | None = None,
    reservation_ttl_seconds: int = 300,
    heartbeat_interval_seconds: float | None = None,
) -> ControlledUsageResult[T]:
    """Call ``AsyncTavilyClient.search`` with a one-credit reservation."""

    if type(query) is not str or not query:
        raise _validation_error()
    options = _options(
        search_options,
        search_depth=search_depth,
        auto_parameters=auto_parameters,
    )
    search = _search_method(client)

    async def invoke() -> T:
        pending = search(query, **options)
        if not inspect.isawaitable(pending):
            raise TypeError("Tavily async client search must return an awaitable")
        return cast(T, await pending)

    return await controlled_usage(
        cost_source_slug=TAVILY_SEARCH_COST_SOURCE_SLUG,
        tool_name=TAVILY_SEARCH_TOOL_NAME,
        metric=TAVILY_SEARCH_METRIC,
        maximum_value=TAVILY_BASIC_SEARCH_CREDITS,
        invoke=invoke,
        extract_actual=_credits,
        customer_id=customer_id,
        step=step,
        reservation_ttl_seconds=reservation_ttl_seconds,
        heartbeat_interval_seconds=heartbeat_interval_seconds,
    )
